File: src/lib/sources/novelfull.ts
const axios = require('axios/dist/browser/axios.cjs');
import { load } from "cheerio";
import { ChapterDataT, ChapterT, NovelT, SOURCES } from "../types";
import { generateEPUB, getPropagandaHTML } from "../epub/epub";
import { NOVEL_CHAPTERS_URI, readChaptersFromInternal, readFileFromInternal, saveChaptersToInternal, saveFileToInternal, saveNovelEpubToDownloads } from "../fs";
import { AutoQueue } from "../queue";
import { UpdateDownloadStatusT } from "../../contexts/DatabaseContext";
import { Dispatch, SetStateAction } from "react";

export const searchNovelFullNovels = async (query: string): Promise<NovelT[]> => {
	try {
		const searchURL = `${SOURCES.NovelFull.url}/search?keyword=${encodeURIComponent(query)}`;
		const res = await axios.get(searchURL);
		const $ = load(res.data);

		const novels: NovelT[] = [];
		$("#list-page .row").each((_, el) => {
			const novelEl = $(el);
			const titleEl = novelEl.find("h3.truyen-title a");
			const url = titleEl.attr("href");
			const title = titleEl.text().trim();
			if (!url || !title) return;
			const author = novelEl.find(".author").text().trim();
			const thumbnailURL = novelEl.find("img").attr("src");
			const novel: NovelT = {
				source: SOURCES.NovelFull.name,
				url: `${SOURCES.NovelFull.url}${url}`,
				title: title,
				authors: author ? [author] : [],
				thumbnailURL: thumbnailURL ? `${SOURCES.NovelFull.url}${thumbnailURL}` : undefined,
			};
			novels.push(novel);
		});

		return novels;
	} catch (error) {
		console.error(error);
		return [];
	}
}

export const loadNovelFullNovel = async (url: string, novelData?: NovelT): Promise<NovelT | undefined> => {
	try {
		const res = await axios.get(url);
		const $ = load(res.data);

		const novelInfoEl = $(".col-info-desc");
		const title = novelInfoEl.find(".desc > h3.title").first().text().trim();
		const rating = novelInfoEl.find(".small").text().trim().replace("Rating: ", "").replace(/from[\S\s]*?(?=\d)/, "from ");
		const description = novelInfoEl.find(".desc-text").text().trim();
		const latestChapter = novelInfoEl.find(".l-chapter > ul.l-chapters > li").first().text().trim();
		const coverURL = novelInfoEl.find(".info-holder .book img").attr("src");
		const authors = novelInfoEl.find(".info-holder > .info > div:first-of-type > a").map((_, el) => $(el).text().trim()).get();
		const alternateTitles = novelInfoEl.find(".info-holder .info > div:nth-of-type(2)").text().replace("Alternative names:", "").trim().split(", ");
		const genres = novelInfoEl.find(".info-holder .info > div:nth-of-type(3) > a").map((_, el) => $(el).text().trim()).get();
		const status = novelInfoEl.find(".info-holder .info > div:nth-of-type(5) > a").text().trim();

		const chaptersEl = $("#list-chapter");
		let chaptersPerPage = 0;
		$(".row ul.list-chapter").each((_, el) => {
			chaptersPerPage += $(el).find("li").length;
		});

		let totalPages = 1;
		let totalChapters = 0;
		const lastPageURL = chaptersEl.find("ul.pagination > li.last").find("a").attr("href");
		if (lastPageURL) {
			// Get total chapters if there are multiple pages
			totalPages = Number.parseInt(lastPageURL.match(/page=(\d+)/)?.[1] || "1");
			totalChapters = chaptersPerPage * (totalPages - 1);

			try {
				const lastPageRes = await axios.get(`${SOURCES.NovelFull.url}${lastPageURL}`);
				const lastPage$ = load(lastPageRes.data);
				lastPage$("#list-chapter .row ul.list-chapter").each((_, el) => {
					totalChapters += $(el).find("li").length;
				});
			} catch (error) {
				console.error("Error while getting total chapters:", error);
			}
		} else {
			// Get total chapters if there is only one page
			totalChapters = chaptersPerPage;
		}

		const novel: NovelT = {
			source: SOURCES.NovelFull.name,
			url,
			title,
			authors,
			genres,
			alternateTitles,
			description,
			coverURL: coverURL ? `${SOURCES.NovelFull.url}${coverURL}` : undefined,
			latestChapter,
			totalChapters,
			status,
			rating,
		};

		return novelData ? { ...novelData, ...novel } : novel;
	} catch (error) {
		console.error(error);
		return novelData;
	}
}

export const downloadNovelFullNovel = async (novel: NovelT, updateDownloadStatus: UpdateDownloadStatusT, isCancel: boolean) => {
	let totalChapters = novel.totalChapters || 0;
	let downloadedChapters = novel.downloadedChapters || 0;

	try {
		console.log(`Downloading novel: ${novel.title}`);
		updateDownloadStatus(novel.url, totalChapters, downloadedChapters, false);

		let res = await axios.get(novel.url);
		let $ = load(res.data);

		const lastPageURL = $("#list-chapter ul.pagination > li.last").find("a").attr("href");
		const totalPages = Number.parseInt(lastPageURL?.match(/page=(\d+)/)?.[1] || "1");

		// Get all chapter links
		console.log(`Getting chapter links from ${totalPages} pages`);
		const chapterLinks: ChapterT[] = [];
		for (let i = 1; i <= totalPages; i++) {
			if (isCancel) throw new Error("Download cancelled");
			console.log(`Getting chapter links from page ${i} of ${totalPages}`);
			$("#list-chapter .row ul.list-chapter > li").each((_, el) => {
				const chapterEl = $(el);
				const title = chapterEl.find("a").text().trim();
				const url = chapterEl.find("a").attr("href");
				if (title && url) chapterLinks.push({ title, url: `${SOURCES.NovelFull.url}${url}` });
				else throw new Error("Failed to get chapter title or url");
			});

			if (i + 1 > totalPages) break;
			res = await axios.get(`${novel.url}?page=${i}`);
			$ = load(res.data);
		}
		totalChapters = chapterLinks.length;
		updateDownloadStatus(novel.url, totalChapters, downloadedChapters, false);
		console.log(`Total chapters: ${totalChapters}`);
		if (isCancel) throw new Error("Download cancelled");

		let chapters: ChapterDataT[] = [];
		// Load downloaded chapters from cache if any
		if (novel.downloadedChapters) {
			const chaptersJsonStr = await readChaptersFromInternal(novel.url);
			if (chaptersJsonStr) chapters = JSON.parse(chaptersJsonStr) as ChapterDataT[];
		}

		// Download chapters
		downloadedChapters = 0;
		const chaptersSaveQueue = new AutoQueue();
		for (let i = 0; i < chapterLinks.length; i++) {
			if (isCancel) throw new Error("Download cancelled");
			const chapter = chapterLinks[i];
			if (chapters.length > i && chapters[i].url === chapter.url && chapters[i].title === chapter.title) {
				downloadedChapters++;
				console.log(`Chapter ${i + 1} of ${chapterLinks.length} already downloaded`);
				continue;
			} else if (chapters.length > i) {
				chapters.splice(i, 1);
			}
			console.log(`Downloading chapter ${i + 1} of ${chapterLinks.length}`);
			const content = await downloadNovelFullChapter(chapter.title, chapter.url);
			if (!content) throw new Error("Failed to download chapter content");
			chapters.push({ ...chapter, content });

			// Cache downloaded chapters to file
			chaptersSaveQueue.enqueue(() => saveChaptersToInternal(novel.url, JSON.stringify(chapters)));
			downloadedChapters++;
			updateDownloadStatus(novel.url, totalChapters, downloadedChapters, false);
		}
		updateDownloadStatus(novel.url, totalChapters, downloadedChapters, false);

		// Generate and save EPUB
		if (isCancel) throw new Error("Download cancelled");
		console.log("Generating EPUB");
		const epubBlob = await generateEPUB(novel, chapters);
		await saveNovelEpubToDownloads(novel.title, epubBlob);
		novel.isDownloaded = true;
	} catch (error) {
		console.error(error);
	}

	novel.totalChapters = totalChapters;
	novel.downloadedChapters = downloadedChapters;
	updateDownloadStatus(novel.url, totalChapters, downloadedChapters, true);
	return novel;
}

const downloadNovelFullChapter = async (title: string, url: string) => {
	try {
		const res = await axios.get(url);
		const $ = load(res.data);

		let contentEl = $("#chapter-content");

		// Remove scripts and iframes
		contentEl.find("script").remove();
		contentEl.find("iframe").remove();

		// Remove ads, classes, ids, styles, comments, etc.
		let content = contentEl.html();
		if (!content) throw new Error("Failed to get chapter content:");
		content = content
			.replace(/class=".*?"/g, "")
			.replace(/id=".*?"/g, "")
			.replace(/style=".*?"/g, "")
			.replace(/data-.*?=".*?"/g, "")
			.replace(/<!--.*?-->/g, "")
			.replace(/<div align="left"[\s\S]*?If you find any errors \( Ads popup, ads redirect, broken links, non-standard content, etc.. \)[\s\S]*?<\/div>/g, "")
			;

		const titleHTML = `<h1>${title}</h1>`;
		const propagandaHTML = getPropagandaHTML();

		content = titleHTML + content + propagandaHTML;

		return content;
	} catch (error) {
		console.error(error);
		return undefined;
	}
}

File: lib/models/sources/novelfull.dart
import 'dart:isolate';

import 'package:html/dom.dart';
import 'package:novelscraper/models/chapter_model.dart';
import 'package:novelscraper/models/novel_isolate.dart';
import 'package:novelscraper/models/novel_model.dart';
import 'package:novelscraper/models/sources/source_model.dart';
import 'package:http/http.dart' as http;
import 'package:html/parser.dart';
import 'package:talker_flutter/talker_flutter.dart';

final talker = TalkerFlutter.init();

class NovelFull {
  static Future<List<Novel>> search(String query) async {
    final List<Novel> novels = [];
    try {
      final searchURI = Uri.https(Source.novelfull.url, "search", {"keyword": query});
      final response = await http.get(searchURI);
      final document = parse(response.body);

      document.querySelectorAll("#list-page .col-truyen-main .row").forEach((novelEl) {
        final titleEl = novelEl.querySelector("h3.truyen-title a");
        if (titleEl == null) return;
        final novelURL = titleEl.attributes["href"];
        final title = titleEl.text.trim();
        if (novelURL == null || title.isEmpty) return;
        final novelURI = Uri.https(Source.novelfull.url, novelURL);

        final author = novelEl.querySelector(".author")?.text.trim() ?? "Unknown";
        final thumbnailURL = novelEl.querySelector("img")?.attributes["src"];

        final novel = Novel(source: Source.novelfull, title: title, url: novelURI.toString());
        novel.authors.add(author);
        novel.thumbnailURL = thumbnailURL != null ? Uri.https(Source.novelfull.url, thumbnailURL).toString() : null;
        novels.add(novel);
      });
    } catch (e, st) {
      talker.handle(e, st, "[NovelFull] Failed to search for novels");
    }
    return novels;
  }

  static Future<Novel?> fetchNovel(Novel novel) async {
    try {
      final novelURI = Uri.parse(novel.url);
      final response = await http.get(novelURI);
      final document = parse(response.body);

      // GET NOVEL INFO
      final novelInfoEl = document.querySelector(".col-info-desc");
      final title = novelInfoEl?.querySelector(".desc > h3.title")?.text.trim();
      final rating = novelInfoEl
          ?.querySelector(".small")
          ?.text
          .trim()
          .replaceFirst("Rating: ", "")
          .replaceFirst(RegExp(r"from[\S\s]*?(?=\d)"), "from ");
      final description = novelInfoEl?.querySelector(".desc-text")?.text.trim();
      final latestChapter = novelInfoEl?.querySelector(".l-chapter > ul.l-chapters > li")?.text.trim();
      final coverURL = novelInfoEl?.querySelector(".info-holder .book img")?.attributes["src"];

      final infoHolderEls = novelInfoEl?.querySelectorAll(".info-holder > .info > div");
      final authors = infoHolderEls?[0].querySelectorAll("a").map((el) => el.text.trim()).toList();
      final alternateTitles = infoHolderEls?[1].text.replaceFirst("Alternative names:", "").trim().split(", ");
      final genres = infoHolderEls?[2].querySelectorAll("a").map((el) => el.text.trim()).toList();
      final status = infoHolderEls?[4].querySelector("a")?.text.trim();

      // GET CHAPTERS PER PAGE
      final chaptersEl = document.querySelector("#list-chapter");
      var chaptersPerPage = 0;
      chaptersEl?.querySelectorAll("ul.list-chapter").forEach((el) {
        chaptersPerPage += el.querySelectorAll("li").length;
      });

      // GET TOTAL CHAPTERS
      var totalPages = 1;
      var totalChapters = 0;
      final lastPageURL = chaptersEl?.querySelector("ul.pagination > li.last")?.querySelector("a")?.attributes["href"];
      if (lastPageURL != null) {
        totalPages = int.tryParse(lastPageURL.split("=").last) ?? 1;
        totalChapters = chaptersPerPage * (totalPages - 1);

        try {
          final lastPageUri = Uri.parse("${novel.url}?page=$totalPages");
          final lastPageRes = await http.get(lastPageUri);
          final lastPageDocument = parse(lastPageRes.body);
          lastPageDocument.querySelectorAll("#list-chapter ul.list-chapter").forEach((el) {
            totalChapters += el.querySelectorAll("li").length;
          });
        } catch (e, st) {
          talker.warning("[NovelFull] Error while getting total chapters", e, st);
        }
      } else {
        totalChapters = chaptersPerPage;
      }

      // UPDATE NOVEL
      novel.title = title ?? novel.title;
      novel.authors = authors ?? novel.authors;
      novel.genres = genres ?? novel.genres;
      novel.alternateTitles = alternateTitles ?? novel.alternateTitles;
      novel.description = description ?? novel.description ?? "No description available.";

      novel.coverURL = coverURL != null ? Uri.https(Source.novelfull.url, coverURL).toString() : null;
      novel.rating = rating ?? novel.rating ?? "No rating available.";
      novel.latestChapterName = latestChapter ?? novel.latestChapterName;
      novel.totalChapters = totalChapters > 0 ? totalChapters : novel.totalChapters;
      novel.status = status ?? novel.status ?? "Unknown";
      return novel;
    } catch (e, st) {
      talker.handle(e, st, "[NovelFull] Failed to fetch novel");
    }
    return null;
  }

  static Future<void> downloadNovel(Map<String, Object> params) async {
    final sPort = params['sPort'] as SendPort;
    final novel = params['novel'] as Novel;
    final preDownloadedChapters = params['preDownloadedChapters'] as List<Chapter>;

    ReceivePort rPort = ReceivePort();
    sPort.send([NovelIsolateAction.setSendPort, rPort.sendPort]);

    // PORT LISTENER TO CANCEL DOWNLOAD IF REQUESTED
    bool cancel = false;
    rPort.listen((message) {
      if (message is List && message.isNotEmpty && message[0] is NovelIsolateAction) {
        switch (message[0]) {
          case NovelIsolateAction.cancel:
            talker.info("[NovelFull] Canceling download");
            cancel = true;
            break;
        }
      }
    });

    try {
      talker.info("[NovelFull] Downloading novel: ${novel.title}");

      final response = await http.get(Uri.parse(novel.url));
      Document document = parse(response.body);

      final lastPageURL = document.querySelector("#list-chapter ul.pagination > li.last")?.querySelector("a")?.attributes["href"];
      final totalPages = lastPageURL != null ? int.tryParse(lastPageURL.split("=").last) ?? 1 : 1;

      const batch = 10; // BATCH OF CHAPTERS TO DOWNLOAD AT ONCE
      List<Chapter> chapters = [];
      for (var i = 1; i <= totalPages; i++) {
        // GET ALL CHAPTERS ON PAGE
        List<Chapter> pageChapters = [];
        document.querySelectorAll("#list-chapter .row ul.list-chapter > li").forEach((chapterEl) {
          final chapterLinkEl = chapterEl.querySelector("a");
          final title = chapterLinkEl?.text.trim();
          final url = chapterLinkEl?.attributes["href"];
          if (title == null && url == null) throw Exception("Chapter title or url not found");
          pageChapters.add(Chapter(title: title!, url: Uri.https(Source.novelfull.url, url!).toString()));
        });
        talker.info("[NovelFull] Downloading chapters: ${chapters.length + 1} to ${chapters.length + pageChapters.length}");

        int skip = 0;
        while (skip < pageChapters.length) {
          // CHECK IF DOWNLOAD IS CANCELED
          if (cancel) throw Exception("Download canceled");

          // CREATE BATCH OF CHAPTERS TO DOWNLOAD
          final batchChapters = pageChapters.skip(skip).take(batch).toList();

          final List<Future<Chapter>> downloadedChapters = [];
          for (int i = 0; i < batchChapters.length; i++) {
            // CHECK IF CHAPTER IS ALREADY DOWNLOADED
            final chapter = batchChapters[i];
            final chaptersIndex = chapters.length + i;
            if (preDownloadedChapters.length > chaptersIndex) {
              final downloadedChapter = preDownloadedChapters[chaptersIndex];
              if (downloadedChapter.content != null &&
                  downloadedChapter.title == chapter.title &&
                  downloadedChapter.url == chapter.url) {
                chapter.content = downloadedChapter.content;
                downloadedChapters.add(Future.value(chapter));
                continue;
              }
            }

            // DOWNLOAD CHAPTER CONTENT IF NOT ALREADY DOWNLOADED
            downloadedChapters.add(_getChapterContent(chapter));
          }

          // WAIT FOR BATCH TO COMPLETE DOWNLOADING
          chapters = [...chapters, ...(await Future.wait(downloadedChapters))];
          novel.downloadedChapters = chapters.length;
          skip += batch;

          // SEND DOWNLOAD PERCENTAGE TO UI
          final percentage = num.parse((((chapters.length) / (novel.totalChapters ?? 0)) * 100).toStringAsFixed(2));
          sPort.send([NovelIsolateAction.setPercentage, percentage]);
        }

        // GET NEXT PAGE IF AVAILABLE
        if (i + 1 > totalPages) break;
        final nextPageURI = Uri.parse("${novel.url}?page=${i + 1}");
        final response = await http.get(nextPageURI);
        document = parse(response.body);
      }
      novel.isDownloaded = true;
      novel.totalChapters = chapters.length;
      talker.info("[NovelFull] Downloaded ${novel.totalChapters} chapters");

      // REQUEST TO GENERATE EPUB
      sPort.send([NovelIsolateAction.generateEPUB, novel, chapters]);
    } catch (e, st) {
      talker.handle(e, st, "[NovelFull] Failed to download novel");
    }

    rPort.close();
    sPort.send([NovelIsolateAction.done, novel]);
  }

  static Future<Chapter> _getChapterContent(Chapter chapter) async {
    final response = await http.get(Uri.parse(chapter.url));
    final document = parse(response.body);

    final contentEl = document.querySelector("#chapter-content");
    if (contentEl == null) throw Exception("Chapter content not found");
    contentEl.querySelectorAll("script").forEach((el) => el.remove());
    contentEl.querySelectorAll("iframe").forEach((el) => el.remove());

    String content = contentEl.innerHtml;
    content = content
        .replaceAll(RegExp(r'class=".*?"'), "")
        .replaceAll(RegExp(r'id=".*?"'), "")
        .replaceAll(RegExp(r'style=".*?"'), "")
        .replaceAll(RegExp(r'data-.*?=".*?"'), "")
        .replaceAll(RegExp(r'<!--.*?-->'), "")
        .replaceAll(
            RegExp(
                r'<div align="left"[\s\S]*?If you find any errors \( Ads popup, ads redirect, broken links, non-standard content, etc.. \)[\s\S]*?<\/div>'),
            "");

    final titleHTML = "<h1>${chapter.title}</h1>";
    final propagandaHTML = Chapter.getPropagandaHTML();

    content = "$titleHTML\n$content\n$propagandaHTML";
    chapter.content = content;
    return chapter;
  }
}
